src/frequenz/sdk/config/_manager.py

```diff
@@ -406,13 +416,56 @@ def __call__(
                     assert False, f"Unexpected match: {unexpected}"
 
         if not has_changed:
-            key_str = f" for key '{self._key}'" if self._key else ""
+            key_str = f" for key '{key}'" if key else ""
             _logger.info("Configuration%s has not changed, skipping update", key_str)
             _logger.debug("Old configuration%s being kept: %r", key_str, old_config)
 
         return has_changed
 
 
+def _get_key(
+    config: Mapping[str, Any],
+    # This is tricky, because a str is also a Sequence[str], if we would use only
+    # Sequence[str], then a regular string would also be accepted and taken as
+    # a sequence, like "key" -> ["k", "e", "y"]. We should never remove the str from
+    # the allowed types without changing Sequence[str] to something more specific,
+    # like list[str] or tuple[str].
+    key: str | Sequence[str] | None,
+) -> Mapping[str, Any] | None:
+    """Get the value from the configuration under the specified key."""
+    if key is None:
+        return config
+    # Order here is very important too, str() needs to come first, otherwise a regular
+    # will also match the Sequence[str] case.
+    # TODO: write tests to validate this works correctly.
+    if isinstance(key, str):
+        key = (key,)
+    value = config
+    current_path = []
+    for subkey in key:
+        current_path.append(subkey)
+        if value is None:
+            return None
+        match value.get(subkey):
+            case None:
+                return None
+            case Mapping() as new_value:
+                value = new_value
+            case _:
+                subkey_str = ""
+                if len(key) > 1:
+                    subkey_str = f" when looking for sub-key {key!r}"
+                _logger.error(
+                    "Found key %r%s but it's not a mapping, returning None: config=%r",
+                    current_path[0] if len(current_path) == 1 else current_path,
+                    subkey_str,
+                    config,
+                )
+                return None
+        value = new_value
+    return value
+
+
```

Comment:
Am I right that this is basically the same but removing `current_path` and using `idx` instead to get it, and removing the `match`?

It is indeed shorter but I find the longer version more clear, at least it took *me* a while to understand the proposed code, I actually like the fact about `match` that makes all the possibilities very clear and separated, but I guess this probably depends on how each brain works, so I can totally get if for you your code is more clear. I'm happy to change it if that's the case, but I'm not so keep if it is just shorter (but not clearer).

Maybe using the index to get a slice instead of creating a new list can be a bit more efficient and more or less as readable, so I could change only that too.

Reply:
Ok , previous is more clear to you, then lets leave it :)